intake/container/base.py

```diff
@@ -1,21 +1,81 @@
-class BaseContainer(object):
+import msgpack
+import requests
+from intake.source.base import DataSource, Schema
+from . import serializer
+from intake import __version__
 
-    @staticmethod
-    def merge(parts):
-        raise NotImplementedError
 
-    @staticmethod
-    def to_dask(parts, dtype):
-        raise NotImplementedError
+class RemoteSource(DataSource):
+    version = __version__
 
-    @staticmethod
-    def encode(obj):
-        raise NotImplementedError
+    def __init__(self, url, headers, name, parameters, metadata=None, **kwargs):
+        super(RemoteSource, self).__init__(self)
+        self.url = url
+        self.name = name
+        self.parameters = parameters
+        self.headers = headers
+        self._source_id = None
+        self.metadata = metadata or {}
+        self._get_source_id()
 
-    @staticmethod
-    def decode(bytestr):
-        raise NotImplementedError
+    def _get_source_id(self):
+        if self._source_id is None:
+            payload = dict(action='open', name=self.name,
+                           parameters=self.parameters)
+            req = requests.post(self.url, data=msgpack.packb(
+                payload, use_bin_type=True), **self.headers)
+            if req.status_code == 200:
+                response = msgpack.unpackb(req.content, encoding='utf-8')
+                self._parse_open_response(response)
 
-    @staticmethod
-    def read(chunks):
-        raise NotImplementedError
\ No newline at end of file
+    def _parse_open_response(self, response):
+        self.datashape = response['datashape']
+        dtype_descr = response['dtype']
+        if isinstance(dtype_descr, list):
+            # Reformat because NumPy needs list of tuples
+            dtype_descr = [tuple(x) for x in response['dtype']]
+        self.dtype = dtype_descr
+        self.shape = tuple(response['shape'])
+        self.npartitions = response['npartitions']
+        self.metadata = response['metadata']
+        self._schema = Schema(datashape=None, dtype=self.dtype,
+                              shape=self.shape,
+                              npartitions=self.npartitions,
+                              extra_metadata=self.metadata)
+        self._source_id = response['source_id']
+
+    def _get_partition(self, i):
+        pass
```

Comment:
todo?

Reply:
I prefer no-op over virtual methods or NotImplemented, just for simplicity.

Reply:
Perhaps a comment then to explain why this is not implemented.

Reply:
Do you personally prefer NotImplemented? That's what's done in the DataSource base class.

Reply:
It depends.  I think the noop is fine if a noop is the default.  If you are depending on the method to actually do something, then I would raise NotImplemented.

Reply:
done